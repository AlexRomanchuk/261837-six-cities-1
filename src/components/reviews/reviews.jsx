import React, {PureComponent} from "react";
import PropTypes from "prop-types";
import {connect} from "react-redux";
import {loadComments} from "../../reducers/reducer.js";
import {format} from "date-fns";

class Reviews extends PureComponent {
  constructor(props) {
    super(props);
  }
  render() {
    const {reviews, isLoaded, isLoading} = this.props;

    if (isLoading) {
      return <h2 className="reviews__title">Loading...</h2>;
    }

    if (!isLoaded) {
      return <h2 className="reviews__title">Error on loading comments.</h2>;
    }

    if (!reviews || !reviews.length) {
      return <h2 className="reviews__title">No reviews</h2>;
    }

    const sortedReviews = reviews.sort((comment1, comment2) => {
      return new Date(comment2.date) - new Date(comment1.date);
    }).slice(0, 10);

    return <div>
      <h2 className="reviews__title">Reviews &middot; <span className="reviews__amount">{reviews.length}</span></h2>
      <ul className="reviews__list">
        {sortedReviews.map((review) => {
          return <li className="reviews__item" key={review.id}>
            <div className="reviews__user user">
              <div className="reviews__avatar-wrapper user__avatar-wrapper">
                <img className="reviews__avatar user__avatar" src={review.user.avatarUrl} width="54" height="54" alt="Reviews avatar" />
              </div>
              <span className="reviews__user-name">
                {review.user.name}
              </span>
            </div>
            <div className="reviews__info">
              <div className="reviews__rating rating">
                <div className="reviews__stars rating__stars">
                  <span style={{width: `${review.rating * 20}%`}}></span>
                  <span className="visually-hidden">Rating</span>
                </div>
              </div>
              <p className="reviews__text">
                {review.comment}
              </p>
              <time className="reviews__time" dateTime={format(new Date(review.date), `YYYY-MM-DD`)}>{format(new Date(review.date), `MMM YYYY`)}</time>
            </div>
          </li>;
        })}
      </ul>
    </div>;
  }
  componentDidMount() {
    const {offerId, load} = this.props;
    load(offerId);
  }
}

Reviews.propTypes = {
  load: PropTypes.func.isRequired,
  offerId: PropTypes.number.isRequired,
  reviews: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number.isRequired,
    rating: PropTypes.number.isRequired,
    comment: PropTypes.string.isRequired,
    date: PropTypes.string.isRequired,
    user: PropTypes.shape({
      id: PropTypes.number.isRequired,
      isPro: PropTypes.bool.isRequired,
      name: PropTypes.string.isRequired,
      avatarUrl: PropTypes.string.isRequired,
    }).isRequired,
  })),
  isLoaded: PropTypes.bool,
  isLoading: PropTypes.bool,
};

const mapStateToProps = (state) => ({
  reviews: state.comments,
  isLoaded: state.isCommentsLoaded,
  isLoading: state.isLoading,
});

const mapDispatchToProps = (dispatch) => ({
  load: (id) => dispatch(loadComments(id))
});

export {Reviews};

export default connect(mapStateToProps, mapDispatchToProps)(Reviews);
