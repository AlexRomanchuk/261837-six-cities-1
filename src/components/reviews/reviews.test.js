import React from "react";
import renderer from "react-test-renderer";
import {Reviews} from "../reviews/reviews.jsx";

const loadedComments = [
  {
    id: 0,
    user: {
      id: 0,
      isPro: true,
      name: `name`,
      avatarUrl: `url`
    },
    rating: 4,
    comment: `some text`,
    date: `2019-05-14T06:07:35.291Z`
  },
  {
    id: 1,
    user: {
      id: 0,
      isPro: true,
      name: `name 2`,
      avatarUrl: `url 2`
    },
    rating: 4,
    comment: `some text 2`,
    date: `2018-05-14T06:07:35.291Z`
  },
];

it(`Reviews snapshot`, () => {
  const tree = renderer.create(<Reviews
    load={jest.fn}
    offerId={1}
    reviews={loadedComments}
    isLoaded={false}
    isLoading={false}
  />).toJSON();
  expect(tree).toMatchSnapshot();
});
