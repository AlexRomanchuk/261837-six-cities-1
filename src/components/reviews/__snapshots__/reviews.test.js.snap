// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reviews snapshot 1`] = `
<h2
  className="reviews__title"
>
  Error on loading comments.
</h2>
`;
